### ISSUE TITLE
Aws::S3::Errors::MalformedXML: The XML you provided was not well-formed or did not validate against our published schema

### ISSUE THREAD

**Issue by mike-bailey** [aws/aws-sdk-ruby#1577]
See: https://github.com/aws/aws-sdk-ruby/issues/1460#issuecomment-322094403

For updates on this issue. 
It's reappeared in a different format (identical to the docs, so either the docs are in some way incorrect or the SDK is doing something wrong) so I figured I'd reopen.

**Comment by awood45** [comment]
Can you provide a wire trace for this?

```ruby
s3 = Aws::S3::Client.new(http_wire_trace: true)
```

**Comment by mike-bailey** [comment]
```
irb(main):014:0> resp = s3.put_bucket_logging({
irb(main):015:2*   bucket: "sourcebucket",
irb(main):016:2*   bucket_logging_status: {
irb(main):017:3*     logging_enabled: {
irb(main):018:4*     target_bucket: "targetbucket",
irb(main):019:4*     },
irb(main):020:3*   },
irb(main):021:2* })
opening connection to sourcebucketname.s3.amazonaws.com:443...
opened
starting SSL for sourcebucketname.s3.amazonaws.com:443...
SSL established
<- "PUT /?logging HTTP/1.1\r\nContent-Type: \r\nAccept-Encoding: \r\nUser-Agent: aws-sdk-ruby2/2.10.25 ruby/2.4.1 x86_64-darwin16\r\nExpect: 100-continue\r\nContent-Md5: gTbnihPA+E+nrr7woTgE5Q==\r\nX-Amz-Date: 20170824T212836Z\r\nHost: sourcebucket.s3.amazonaws.com\r\nX-Amz-Content-Sha256: 7906216862bfdea0bc9a4dd1182b4315bbd2dc6ce8884259a1101a46dbdfd05e\r\nAuthorization: AWS4-HMAC-SHA256 Credential=redact/20170824/us-east-1/s3/aws4_request, SignedHeaders=content-md5;host;x-amz-content-sha256;x-amz-date, Signature=redact\r\nContent-Length: 202\r\nAccept: */*\r\n\r\n"
-> "HTTP/1.1 100 Continue\r\n"
-> "\r\n"
-> "HTTP/1.1 400 Bad Request\r\n"
-> "x-amz-request-id: 8D18B1FA427DF7C5\r\n"
-> "x-amz-id-2: redacted+redacted/6AhyISj4Q=\r\n"
-> "Content-Type: application/xml\r\n"
-> "Transfer-Encoding: chunked\r\n"
-> "Date: Thu, 24 Aug 2017 21:28:30 GMT\r\n"
-> "Connection: close\r\n"
-> "Server: AmazonS3\r\n"
-> "\r\n"
-> "13f\r\n"
reading 319 bytes...
-> "<?xml version=\"1.0\" encoding=\"UTF-8\"?>\n<Error><Code>MalformedXML</Code><Message>The XML you provided was not well-formed or did not validate against our published schema</Message><RequestId>8D18B1FA427DF7C5</RequestId><HostId>SqGmM48esm+NqvmkVTKvbFRBche3itT0uuuHfsOy8iyx5HVkDx6Qqb6nqs0NcRkkk/6AhyISj4Q=</HostId></Error>"
read 319 bytes
reading 2 bytes...
-> "\r\n"
read 2 bytes
-> "0\r\n"
-> "\r\n"
Conn close
Aws::S3::Errors::MalformedXML: The XML you provided was not well-formed or did not validate against our published schema
	from /usr/local/lib/ruby/gems/2.4.0/gems/aws-sdk-core-2.10.25/lib/seahorse/client/plugins/raise_response_errors.rb:15:in `call'
	from /usr/local/lib/ruby/gems/2.4.0/gems/aws-sdk-core-2.10.25/lib/aws-sdk-core/plugins/s3_sse_cpk.rb:19:in `call'
	from /usr/local/lib/ruby/gems/2.4.0/gems/aws-sdk-core-2.10.25/lib/aws-sdk-core/plugins/s3_dualstack.rb:24:in `call'
	from /usr/local/lib/ruby/gems/2.4.0/gems/aws-sdk-core-2.10.25/lib/aws-sdk-core/plugins/s3_accelerate.rb:34:in `call'
	from /usr/local/lib/ruby/gems/2.4.0/gems/aws-sdk-core-2.10.25/lib/aws-sdk-core/plugins/jsonvalue_converter.rb:20:in `call'
	from /usr/local/lib/ruby/gems/2.4.0/gems/aws-sdk-core-2.10.25/lib/aws-sdk-core/plugins/idempotency_token.rb:18:in `call'
	from /usr/local/lib/ruby/gems/2.4.0/gems/aws-sdk-core-2.10.25/lib/aws-sdk-core/plugins/param_converter.rb:20:in `call'
	from /usr/local/lib/ruby/gems/2.4.0/gems/aws-sdk-core-2.10.25/lib/seahorse/client/plugins/response_target.rb:21:in `call'
	from /usr/local/lib/ruby/gems/2.4.0/gems/aws-sdk-core-2.10.25/lib/seahorse/client/request.rb:70:in `send_request'
	from /usr/local/lib/ruby/gems/2.4.0/gems/aws-sdk-core-2.10.25/lib/seahorse/client/base.rb:207:in `block (2 levels) in define_operation_methods'
	from (irb):14
	from /usr/local/bin/irb:11:in `<main>'
irb(main):022:0> 
```
I removed anything I thought could have been private. If I'm wrong about that please LMK.

**Comment by awood45** [comment]
Thanks, and I've been able to recreate this issue as well. Investigating whether this is a client or service issue, since I don't think we've touched the relevant code in some time.

**Comment by mike-bailey** [comment]
Thank you! Great response time.

**Comment by awood45** [comment]
I've reached out to the service team for help debugging this as well.

**Comment by mike-bailey** [comment]
Cool, thanks. If it matters at all there's a `-` in one of the bucket names (which is mandatory per our use case). Let me know if I can help further.

**Comment by awood45** [comment]
Thanks for the note - I don't think that's a factor since I reproduced the error without a dash.

**Comment by mike-bailey** [comment]
Hello, any updates yet?

**Comment by mike-bailey** [comment]
Any news? 😄

**Comment by ddimmich** [comment]
having the same issue from within logstash.  would be great to get an update.

**Comment by mike-bailey** [comment]
Anyone home?

**Comment by awood45** [comment]
Sorry, no updates yet. Will revisit after reInvent, need to collaborate with a few parties on this one.

**Comment by mike-bailey** [comment]
No problem, just making sure it wasn't abandoned.

**Comment by mike-bailey** [comment]
@cjyclaire Any updates regarding this? What's the expected turnaround if you don't mind my asking?

**Comment by cjyclaire** [comment]
Forward the ping again : ) Appreciate your patience
Also, feel free to cut a TT to AWS Support at the same time as well, they can help look into the request IDs.

**Comment by mike-bailey** [comment]
Unfortunately I can't open a technical ticket as we don't pay for that level of support.

**Comment by awood45** [comment]
May have a breakthrough here, as it appears S3 sometimes throws these errors for general usage mistakes, rather than a malformed XML request. Checking.

**Comment by awood45** [comment]
Potentially related Issue #1670

I've passed this information on to the S3 team and will look for a new update.

**Comment by mike-bailey** [comment]
Interesting. Thanks for the update. Neither of these are supposed to be improper calls though, no?
Lifecycle policies should be allowed to apply to buckets and logging as well?

**Comment by awood45** [comment]
To be entirely honest, I'm not sure. The documentation doesn't indicate that the call is improper, but I can't eliminate that possibility either. As far as I can tell:

* The Ruby SDK is properly formatting your request.
* This same error has appeared elsewhere in the case of user error, though of a different kind.

**Comment by admorphit** [comment]
This is happening for the nodejs aws-sdk as well in case that helps.

**Comment by awood45** [comment]
@admorphit That's actually very helpful. Do you have a related issue, or a JS code snippet to reproduce that you can provide? Thanks!

**Comment by mike-bailey** [comment]
That also helps me because it gives me something to tell our management about how it’s more than likely not the Ruby SDK’s fault (other than @awood45 having just said it it isn’t). Thanks!

**Comment by awood45** [comment]
Okay, I've identified the root cause. As suspected, it's user error, though granted, the requirements here are not well documented (which is feedback I'm pressing with the service team).

* The `:target_prefix` parameter is required. This is why you're getting an error.
* You'll also likely need to set grant permissions, but at least for that you'll get a much more user-friendly error and the way to do that is documented.

So, it's again much like the other issue, where MalformedXML is being returned in place of a user-friendly error message. This should unblock you - closing, but feel free to add any follow-up questions you have.

**Comment by mike-bailey** [comment]
>The :target_prefix parameter is required. This is why you're getting an error.


However, in the console, a prefix isn't required? Why is this an API limitation only for these SDKs and not the console? I'd like to be able to apply it to the entire bucket...

**Comment by mike-bailey** [comment]
Also, any developments on the logging call? (I'm assuming you're referring to the lifecycle bit). Thank you!

**Comment by mike-bailey** [comment]
Looks like in the actual API:

https://docs.aws.amazon.com/AmazonS3/latest/dev/lifecycle-configuration-examples.html

>If you want the lifecycle rule to apply to all objects in the bucket, specify an empty prefix. 

Maybe something to add to the docs for the actual SDKs. Thank you!

**Comment by awood45** [comment]
I agree with you entirely - documentation across the SDKs comes from the service teams, and I'm pushing for these updates across the impacted APIs currently. Once done, they'll be included in the standard release process.

**Comment by awood45** [comment]
> However, in the console, a prefix isn't required? Why is this an API limitation only for these SDKs and not the console? I'd like to be able to apply it to the entire bucket...

It is required. My guess is that leaving the field blank is translating to an empty target prefix, which is valid. A null target prefix is not valid. This should be better documented.

**Comment by mike-bailey** [comment]
>My guess is that leaving the field blank is translating to an empty target prefix

Thank you. That makes perfect sense in hindsight.

Still learning as a measly intern, so I appreciate the patience. Thank you!

**Comment by awood45** [comment]
No worries, this is a very unhelpful error message so your frustration is understood and appreciated. We're pushing for a better user experience here.

**Comment by ddimmich** [comment]
In the context of logstash which uses this library - how would one add the target prefix parameter?  Is this something that needs to be addressed in logstash or in this library?  Thank you

**Comment by awood45** [comment]
That would be added in the caller - logstash. Unless logstash plumbs through your own parameters, in which case you would need to add it there.

**Comment by ddimmich** [comment]
Ok - Thanks.  Not sure this is one of the parameters that would get 
passed through - haven't seen a configuration equivalent so it sounds 
like this would need to be fixed upstream :/

**Comment by ramd123** [comment]
I was facing this issue while setting analytics configuration using aws java sdk, this should have fixed Amazon by now. any solution to skip filter/predicate while creating analytics using java sdk?

**Comment by mike-bailey** [comment]
Did you do a blank one like @awood45 suggests in this issue?

**Comment by ramd123** [comment]
no, blank also not working..  same exception with blank as well.

**Comment by mike-bailey** [comment]
Maybe post the code if you can?

**Comment by ramd123** [comment]
private void manageAnalytics(String name, AmazonS3 s3) {
		System.out.println("bucket name => " + name);
		if (name.equalsIgnoreCase("genie-files")) {
			SetBucketAnalyticsConfigurationRequest req = new SetBucketAnalyticsConfigurationRequest();
			req.setBucketName(name);
			AnalyticsS3BucketDestination s3destination = new AnalyticsS3BucketDestination()
					.withBucketAccountId("12345").withBucketArn("arn:aws:s3:::test-lifecycle")
					.withFormat(AnalyticsS3ExportFileFormat.CSV);
			AnalyticsExportDestination destination = new AnalyticsExportDestination()
					.withS3BucketDestination(s3destination);
			StorageClassAnalysisDataExport dataExport = new StorageClassAnalysisDataExport()
					.withDestination(destination).withOutputSchemaVersion("V_1");
			AnalyticsConfiguration analyticsConfiguration = new AnalyticsConfiguration()
					.withId(name + "-analytics-conf-2")
					.withStorageClassAnalysis(new StorageClassAnalysis().withDataExport(dataExport))
					.withFilter(new AnalyticsFilter().withPredicate(new AnalyticsPrefixPredicate("123456")));
			req.setAnalyticsConfiguration(analyticsConfiguration);
			System.out.println("before setting analytics configuration => " + req.toString());
			s3.setBucketAnalyticsConfiguration(req);
			System.out.println("after setting analytics configuration => ");
		}
	}

**Comment by ramd123** [comment]
I see a similar exception when i create inventory configuration with optional fields, not sure how come Amazon tested this? without optional fields it works, but with optional fields it was failing with same exception. The error messaging should be corrected, it should be user friendly

SetBucketInventoryConfigurationRequest req = new SetBucketInventoryConfigurationRequest();
			req.setBucketName(name);
			InventoryS3BucketDestination s3BucketDestination = new InventoryS3BucketDestination()
					.withBucketArn("arn:aws:s3:::test-lifecycle").withAccountId("1234567")
					.withFormat(InventoryFormat.CSV);
			InventoryDestination destination = new InventoryDestination().withS3BucketDestination(s3BucketDestination);
			List<String> optionalFields = new ArrayList<String>(Arrays.asList("Size", "LastModifiedDate", "StorageClass", "Etag", "IsMultipartUploaded", "ReplicationStatus", "EncryptionStatus"));
			InventoryConfiguration inventoryConfiguration = new InventoryConfiguration().withDestination(destination)
					.withId(name + "-inventory-conf-1").withEnabled(Boolean.TRUE)
					.withIncludedObjectVersions(InventoryIncludedObjectVersions.Current)
					.withSchedule(new InventorySchedule().withFrequency(InventoryFrequency.Daily)).withOptionalFields(optionalFields);
			req.setInventoryConfiguration(inventoryConfiguration);
			s3.setBucketInventoryConfiguration(req);

**Comment by awood45** [comment]
@ramd123 I'd recommend opening an issue on the [AWS SDK for Java repo](https://github.com/aws/aws-sdk-java) - hard for me to debug any edge cases you may be running in to with that tool.

**Comment by ramd123** [comment]
Thanks, I have created the Issue - https://github.com/aws/aws-sdk-java/issues/1613

**Comment by eric-swann-q2** [comment]
Thanks be to this thread!  This issue also cropped up for me when setting up bucket replication.  The lack of a prefix causes this exception as well, which is super unhelpful.  Thanks!

**Comment by pcchannz** [comment]
@eric-swann-q2 I have the same the problem for setting bucket replication too but via c# sdk. The workaround i have is to add empty filter, priority and DeleteMarkerReplication.

**Comment by rtrscience** [comment]
> @eric-swann-q2 I have the same the problem for setting bucket replication too but via c# sdk. The workaround i have is to add empty filter, priority and DeleteMarkerReplication.

Can you please elaborate the work around for the c# version of this error..

**Comment by mullermp** [comment]
@rtrscience If you are having trouble, feel free to open an issue on their SDK: https://github.com/aws/aws-sdk-net

**Comment by odykyi** [comment]
I have the same issue 

```xml
<?xml version="1.0" encoding="UTF-8"?>
<Error>
    <Code>MalformedXML</Code>
    <Message>The XML you provided was not well-formed or did not validate against our published schema</Message>
    <RequestId>9E938D6616E4B531</RequestId>
    <HostId>3iTCg2o9sft8GxEB72xB9th9kFbphYKeztV9xqc+hpQsI1qqfEETwKV5ECfSIjDxt7wlI8mHVvo=</HostId>
</Error>
```

**Comment by NeelAurea** [comment]
The recommendation mentioned [here ](https://github.com/aws/aws-sdk-ruby/issues/1577#issuecomment-355111872) worked for me. I am attempting to assign transition policies to the buckets using Python SDK. Using an empty string for 'Prefix' in the 'Filter' section did the trick.

**Comment by JovanToroman** [comment]
For me, this issue happened on the `DeleteObjects` command and it happened because the keys I was trying to delete contained escapeable html characters (`&`, `<`, `>`). After replacing these characters with their object versions (`&amp;`, `&lt;`, `&gt;`), everything worked as expected.

**Comment by yaminir** [comment]
@awood45 I still see the issue even after inserting a prefix in my boto3api call for put_bucket_replication. Here is is my error: botocore.exceptions.ClientError: An error occurred (MalformedXML) when calling the PutBucketReplication operation: The XML you provided was not well-formed or did not validate against our published schema. Could you please let me know what else could be done? I am writing my lambda in Python to create a custom resource

**Comment by alextwoods** [comment]
@yaminir  - I'd recommend creating an issue in the [Boto3 repo](https://github.com/boto/boto3).

**Comment by JoeyJGarcia** [comment]
I got this message too.  I am coding in Node. This error happens when trying to upload a PDF document in the size range of 20MB, 10MB works fine, but somewhere between 10MB & 20MB there is a value to make it break.  Also, we tried combining two 10MB PDF and it fails, so it is definitely a size issue in our situation.

Line of code making the call:       await this.s3Connector.upload(putParams);
Paramater:       const putParams = {
        Bucket: process.env.DOC_BUCKET_NAME,
        Key,
        Body: readStream,
      };
      
Error Message: MalformedXML: The XML you provided was not well-formed or did not validate against our published schema
